src/main/java/net/minecraftforge/common/extensions/IForgeFriendlyByteBuf.java

```diff
@@ -152,4 +158,25 @@ default FluidStack readFluidStack()
     {
         return !self().readBoolean() ? FluidStack.EMPTY : FluidStack.readFromPacket(self());
     }
+    
+    /**
+     * Read a Collection from this buffer with set size.
+     */
+    default <T, C extends Collection<T>> C readCollectionWithSize(int size, IntFunction<C> ctor, Function<FriendlyByteBuf, T> builder)
+    {
+        C c = ctor.apply(size);
+
+        for (int j = 0; j < size; ++j) {
+            c.add(builder.apply(self()));
+        }
+        return c;
+    }
+
+    /**
+     * Read a List from this buffer with set size.
+     */
+    default <T> List<T> readListWithSize(int size, Function<FriendlyByteBuf, T> builder)
+    {
+        return this.readCollectionWithSize(size, Lists::newArrayListWithCapacity, builder);
```

Comment:
Can I point out, the following javadoc on `Lists#newArrayListWithCapacity`
<b>Note for Java 7 and later:</b> this method is now unnecessary and should be treated as deprecated. Instead, use `new ArrayList(int)` directly, taking advantage of the new <a href="http://goo.gl/iz2Wi">"diamond" syntax</a>.

*so... why write `Lists::newArrayListWithCapacity` when you could write `ArrayList::new`...*

Reply:
This is what vanilla used for `FriendlyByteBuf#readCollection` I kept it the same for consistency.

Reply:
Whats wrong with doing the limit that we did in the old patch and not needing any of this at all?

Reply:
TBH I didn't checked, what the 1.16 patch was. I thought this patch was omitted because of vanilla changes.

Reply:
Always check, the people who update this are, unfortunately, human. And things sometimes get missed.

Reply:
So I checked and the limit was vanilla behavior that was changed to use the new `FriendlyByteBuf#readList`. It should still works though.